File: leetcode/problems/53.maximum-subarray/53.maximum-subarray.go
package main

func maxSubArray(nums []int) int {
	if len(nums) <= 0 {
		return 0
	}
	if len(nums) == 1 {
		return nums[0]
	}

	largestSum, intermediateSum := nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		intermediateSum = findMax(nums[i], nums[i]+intermediateSum)
		if intermediateSum > largestSum {
			largestSum = intermediateSum
		}
	}

	return largestSum
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: leetcode/problems/problem-specific/53.maximum-subarray/53.maximum-subarray.py
#
# @lc app=leetcode id=53 lang=python3
#
# [53] Maximum Subarray
#

# @lc code=start

# * Kadane's Algorithm Solution | O(n) Time | O(1) Space
# * n -> The length of nums array


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_sum = nums[0]
        cur_sum = 0
        for num in nums:
            # * Start the subarray from the current num (num) or extend the
            # * subarray using the current num (num + cur_sum).
            cur_sum = max(num, num + cur_sum)
            max_sum = max(max_sum, cur_sum)

        return max_sum


# @lc code=end
